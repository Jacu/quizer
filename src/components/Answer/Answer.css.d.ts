export const Answer: string;
export const selected: string;
export const correct: string;
export const incorrect: string;
