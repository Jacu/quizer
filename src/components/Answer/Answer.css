.Answer {
    display: inline-block;
    width: 50%;        
    /* border: 1px solid rgba(0,0,0,0.1); */
    padding: 15px;
    cursor: pointer;
}

.selected {
    background: #125271;
    color: white; 
}

.correct {
    background: green;
    color: white;
}

.incorrect {
    background: red;
    color: white;
}

.Answer:hover {
    background: #b3d8e9; 
}