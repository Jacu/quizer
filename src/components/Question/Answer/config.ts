enum AnswerStates { CorrectSelected = 'CS', IncorrectSelected = 'IS', Selected = 'S', Notselected = 'NS' };

export {
    AnswerStates,
}