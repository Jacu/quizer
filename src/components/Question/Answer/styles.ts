import styled from 'styled-components';
import { CORRECT, WRONG, GRAY, PRIMARY_THEME } from '~/utils/colors';

interface IAnswerProps {
    selected: boolean,
    showAnswer: boolean,
    correct: boolean,
}

enum StyleState { CorrectSelected, IncorrectSelected, Selected, Notselected }

const getState = (props: IAnswerProps): StyleState => {
    const { selected, showAnswer, correct } = props;
    if (selected) {
        if (showAnswer) {
            if(correct) {
                return StyleState.CorrectSelected;
            } else {
                return StyleState.IncorrectSelected;
            }
        } else {
            return StyleState.Selected;
        }
    }
    return StyleState.Notselected;    
};

const answerBackgroundColor = {
    [StyleState.CorrectSelected]: CORRECT.LIGHT,
    [StyleState.IncorrectSelected]: WRONG.LIGHT,
    [StyleState.Selected]: 'transparent',
    [StyleState.Notselected]: 'transparent',
}

const answerBorderColor = {
    [StyleState.CorrectSelected]: CORRECT.BASE,
    [StyleState.IncorrectSelected]: WRONG.BASE,
    [StyleState.Selected]: GRAY.BASE,
    [StyleState.Notselected]: GRAY.BASE,
}

export const Answer = styled.div<IAnswerProps>`
    display: flex;
    width: 100%;
    padding: 15px 0;
    cursor: pointer;
    background-color: ${props => answerBackgroundColor[getState(props)]};
    :hover {
        background: ${GRAY.LIGHT};
    }
    border-width: 1px;
    border-style: solid;
    border-color: ${props => answerBorderColor[getState(props)]};
    margin: 5px 0;
    border-radius: 5px;
`;

const checkboxBorderColor = {
    [StyleState.CorrectSelected]: CORRECT.BASE,
    [StyleState.IncorrectSelected]: WRONG.BASE,
    [StyleState.Selected]: PRIMARY_THEME.BASE,
    [StyleState.Notselected]: GRAY.BASE,
}

const checkboxBorderWidth = {
    [StyleState.CorrectSelected]: '10px',
    [StyleState.IncorrectSelected]: '6px',
    [StyleState.Selected]: '6px',
    [StyleState.Notselected]: '1px',
}

export const Checkbox = styled.div<IAnswerProps>`
    width: 20px;
    height: 20px;
    border-width: ${props => checkboxBorderWidth[getState(props)]};
    border-style: solid;
    border-color: ${props => checkboxBorderColor[getState(props)]};
    margin: 0 10px;
    border-radius: 50%;
`;