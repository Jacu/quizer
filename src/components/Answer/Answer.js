import React from 'react';
import styles from './Answer.css'

const Answer = ({selected,onClick,answer,correct,ended}) => {    
    const className = [styles.Answer,selected ? ended ? correct ? styles.correct : styles.incorrect : styles.selected : ended ? correct ? styles.correct : '' : ''];
    return (
    <div 
        className={className.join(' ')}
        onClick={ended ? null : onClick}>
        {answer}
    </div>)
};

export default Answer;