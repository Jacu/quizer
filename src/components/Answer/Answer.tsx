import React from 'react';
import * as styled from './styles';

interface AnswerProps {
    selected: boolean;
    answer: string;
    correct: boolean;
    ended: boolean;
    onClick: () => void;
}

const Answer: React.FC<AnswerProps> = ({selected, onClick, answer, correct, ended}) => {
    const handleClick = () => {
        if(!ended) onClick();
    };

    return (
    <styled.Answer
        children={answer}
        isSelected={selected}
        isCorrect={correct}
        showAnswer={ended}
        onClick={handleClick} />
    );
};

export default Answer;