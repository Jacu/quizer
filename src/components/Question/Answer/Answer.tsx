import React from 'react';
import * as styled from './styles';

interface AnswerProps {
    answer: string,
    isSelected: boolean,
    isCorrect: boolean,
    reveal: boolean,
    onClick: () => void,
}

const Answer: React.FC<AnswerProps> = ({isSelected, onClick, answer, isCorrect, reveal}) => 
    <styled.Answer 
        isSelected={isSelected}
        isCorrect={isCorrect}
        showAnswer={reveal} 
        onClick={!reveal ? onClick : undefined} >
        <styled.Checkbox showAnswer={reveal} isCorrect={isCorrect} isSelected={isSelected} />
        {answer} 
    </styled.Answer>

export default Answer;