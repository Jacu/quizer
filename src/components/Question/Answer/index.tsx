import React from 'react';
import * as styled from './styles';

interface IAnswerProps {
    answer: string,
    selected: boolean,
    correct: boolean,
    reveal: boolean,
    onClick: () => void,
}

const Answer: React.FC<IAnswerProps> = ({ selected, onClick, answer, correct, reveal }) =>
    <styled.Answer
        selected={selected}
        correct={correct}
        showAnswer={reveal}
        onClick={!reveal ? onClick : undefined} >
        <styled.Checkbox showAnswer={reveal} correct={correct} selected={selected} />
        {answer}
    </styled.Answer>

export default Answer;