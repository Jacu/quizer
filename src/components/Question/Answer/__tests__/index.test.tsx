import React from 'react';
import Answer from '../';
import { render, fireEvent } from '@testing-library/react';

const onClick = jest.fn();
const answer = "answer text";

const props = {
    answer,
    selected: false,
    correct: false,
    reveal: false,
    onClick,
}

describe('SettingPanel', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('should render with default props', () => {
        const { asFragment } = render(<Answer {...props} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('should render when selected', () => {
        const { asFragment } = render(<Answer {...props} selected={true} />);
        expect(asFragment()).toMatchSnapshot();
        expect(asFragment()).toEqual(render(<Answer {...props} selected={true} reveal={false} correct={false} />).asFragment());
        expect(asFragment()).toEqual(render(<Answer {...props} selected={true} reveal={false} correct={true} />).asFragment());
    });

    it('should render when selected and reveal and correct', () => {
        const { asFragment } = render(<Answer {...props} selected={true} reveal={true} correct={true} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('should render when selected and reveal and incorrect', () => {
        const { asFragment } = render(<Answer {...props} selected={true} reveal={true} correct={false} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('should render unselected answer',() => {
        const selected = false;
        const { asFragment } = render(<Answer {...props} selected={selected} />);
        expect(asFragment()).toMatchSnapshot();
        expect(asFragment()).toEqual(render(<Answer {...props} selected={selected} reveal={true} correct={false} />).asFragment());
        expect(asFragment()).toEqual(render(<Answer {...props} selected={selected} reveal={false} correct={true} />).asFragment());
        expect(asFragment()).toEqual(render(<Answer {...props} selected={selected} reveal={true} correct={true} />).asFragment());
        expect(asFragment()).toEqual(render(<Answer {...props} selected={selected} reveal={false} correct={false} />).asFragment());
    });

    it('should call onClick when not in reveal mode', () => {
        const { getByText } = render(<Answer {...props} />);
        fireEvent.click(getByText(answer));
        expect(onClick).toHaveBeenCalledTimes(1);
    });

    it('should NOT call onClick when in reveal mode', () => {
        const { getByText } = render(<Answer {...props} reveal={true} />);
        fireEvent.click(getByText(answer));
        expect(onClick).not.toHaveBeenCalled();
    });
});