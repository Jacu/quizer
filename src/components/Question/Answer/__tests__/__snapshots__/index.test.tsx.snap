// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingPanel should render when selected 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm Apjit"
  >
    <div
      class="sc-AxirZ eKayzj"
    />
    answer text
  </div>
</DocumentFragment>
`;

exports[`SettingPanel should render when selected and reveal and correct 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm kcwPPb"
  >
    <div
      class="sc-AxirZ BHBvY"
    />
    answer text
  </div>
</DocumentFragment>
`;

exports[`SettingPanel should render when selected and reveal and incorrect 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm iLpmua"
  >
    <div
      class="sc-AxirZ eRsxqP"
    />
    answer text
  </div>
</DocumentFragment>
`;

exports[`SettingPanel should render with default props 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm Apjit"
  >
    <div
      class="sc-AxirZ gphSZZ"
    />
    answer text
  </div>
</DocumentFragment>
`;
